python/pyspark/sql/connect/functions/builtin.py

```diff
@@ -2476,8 +2476,26 @@ def repeat(col: "ColumnOrName", n: Union["ColumnOrName", int]) -> Column:
 repeat.__doc__ = pysparkfuncs.repeat.__doc__
 
 
-def split(str: "ColumnOrName", pattern: str, limit: int = -1) -> Column:
-    return _invoke_function("split", _to_col(str), lit(pattern), lit(limit))
+def split(
+    str: "ColumnOrName",
+    pattern: Union[Column, str],
+    limit: Union["ColumnOrName", int] = -1,
+) -> Column:
+    # work around shadowing of str in the input variable name
+    from builtins import str as py_str
+
+    if isinstance(pattern, py_str):
+        _pattern = lit(pattern)
+    elif isinstance(pattern, Column):
+        _pattern = pattern
+    else:
+        raise PySparkTypeError(
+            error_class="NOT_COLUMN_OR_STR",
+            message_parameters={"arg_name": "pattern", "arg_type": type(pattern).__name__},
+        )
+
+    limit = lit(limit) if isinstance(limit, int) else _to_col(limit)
+    return _invoke_function("split", _to_col(str), _pattern, limit)
```

Comment:
```suggestion
    limit = lit(limit) if isinstance(limit, int) else _to_col(limit)
    return _invoke_function("split", _to_col(str), lit(pattern), limit)
```

Reply:
Thanks for the suggestion, this is simpler for sure! The only concern is that we will not raise `PySparkTypeError` if `pattern` is passed in for a type other than `Column` or `str`, and it will form a UnresolvedFunction. Is raising such error early a requirement for connect?

Reply:
Only a few functions have such check, and most functions don't check the types.
We might need to figure out an easy way for type checking.

As to this function, let's keep it simpler for now.

Reply:
Sure. I removed the type check now. Maybe in the future we can standard this, e.g. with a decorator that inspect the function signature and do type check accordingly.

Reply:
Sure. I removed the type check now. Maybe in the future we can standardize this, e.g. with a decorator that inspect the function signature and do type check accordingly.